content/en/blog/_posts/2024-05-01-cri-streaming-explained.md

```diff
@@ -0,0 +1,338 @@
+
+Interpreting data from for the defined protocol is fairly simple: The first
+byte of every input and output packet [defines](https://github.com/kubernetes/kubernetes/blob/9791f0d/staging/src/k8s.io/apimachinery/pkg/util/remotecommand/constants.go#L57-L64)
+the actual stream:
+
+| First Byte | Type            | Description                              |
+| ---------- | --------------- | ---------------------------------------- |
+| `0`        | standard input  | Data streamed from stdin                 |
+| `1`        | standard output | Data streamed to stdout                  |
+| `2`        | standard error  | Data streamed to stderr                  |
+| `3`        | stream error    | A streaming error occurred               |
+| `4`        | stream resize   | A terminal resize event                  |
+| `255`      | stream close    | Stream should be closed (for WebSockets) |
+
+How should runtimes now implement the streaming server methods for **Exec** and
+**Attach** by using the provided kubelet library? The key is that the streaming
+server implementation in the kubelet [outlines an interface](https://github.com/kubernetes/kubernetes/blob/db9fcfe/staging/src/k8s.io/kubelet/pkg/cri/streaming/server.go#L63-L68)
+called `Runtime` which has to be fulfilled by the actual container runtime if it
+wants to use that library:
+
+```go
+// Runtime is the interface to execute the commands and provide the streams.
+type Runtime interface {
+        Exec(ctx context.Context, containerID string, cmd []string, in io.Reader, out, err io.WriteCloser, tty bool, resize <-chan remotecommand.TerminalSize) error
+        Attach(ctx context.Context, containerID string, in io.Reader, out, err io.WriteCloser, tty bool, resize <-chan remotecommand.TerminalSize) error
+        PortForward(ctx context.Context, podSandboxID string, port int32, stream io.ReadWriteCloser) error
+}
+```
+
+Everything related to the protocol interpretation is
+already in place and runtimes only have to implement the actual `Exec` and
+`Attach` logic. For example the container runtime [CRI-O](https://github.com/cri-o/cri-o)
+does it [like this pseudo code](https://github.com/cri-o/cri-o/blob/2a0867/server/container_exec.go#L27-L46):
+
+```go
+func (s StreamService) Exec(
+    ctx context.Context,
+    containerID string,
+    cmd []string,
+    stdin io.Reader, stdout, stderr io.WriteCloser,
+    tty bool,
+    resizeChan <-chan remotecommand.TerminalSize,
+) error {
+    // Retrieve the container by the provided containerID
+    // …
+
+    // Update the container status and verify that the workload is running
+    // …
+
+    // Execute the command and stream the data
+    return s.runtimeServer.Runtime().ExecContainer(
+        s.ctx, c, cmd, stdin, stdout, stderr, tty, resizeChan,
+    )
+}
+```
+
+### **PortForward**
+
+Forwarding ports to a container works a bit differently when comparing it to
+streaming IO data from a workload. The server still has to provide an URL
```

Comment:
```suggestion
streaming IO data from a workload. The server still has to provide a URL
```

Reply:
I've had this question in the past, and even though the source isn't what you would consider an academic reference, there are some good explanations here: https://www.quora.com/Which-is-correct-A-URL-or-An-URL